docs/design/queue-level-scheduling-policy.md

```diff
@@ -0,0 +1,289 @@
+
+```go
+func OpenSession(cache cache.Cache, tiers []conf.Tier, configurations []conf.Configuration) *Session {
+	ssn := openSession(cache)
+	ssn.Tiers = tiers
+	ssn.Configurations = configurations
+	ssn.NodeMap = GenerateNodeMapAndSlice(ssn.Nodes)
+	ssn.PodLister = NewPodLister(ssn)
+
+  // Global plugin registration is still retained.
+	for _, tier := range tiers {
+		for _, plugin := range tier.Plugins {
+			if pb, found := GetPluginBuilder(plugin.Name); !found {
+				klog.Errorf("Failed to get plugin %s.", plugin.Name)
+			} else {
+				plugin := pb(plugin.Arguments)
+				ssn.plugins[plugin.Name()] = plugin
+				onSessionOpenStart := time.Now()
+				plugin.OnSessionOpen(ssn)
+				metrics.UpdatePluginDuration(plugin.Name(), metrics.OnSessionOpen, metrics.Duration(onSessionOpenStart))
+			}
+		}
+	}
+  
+  // A new queue-level plugin registration is added.
+    ssn.QueueContexts = make(map[api.QueueID]*QueueContext)
+
+    // Traverse all queues.
+	for _, queue := range cache.GetQueues() {
+		qCtx := &QueueContext{
+			Plugins:     []Plugin    
+			NodeOrderFns: make(map[string]NodeOrderFn),
+			...
+		}
+        
+        // Process queue-level plugins.
+        if Plugins := queue.Spec.Plugins; Plugins != nil {
+            for _, plugin := range Plugins {
+                pb _ := GetPluginBuilder(plugin.Name)
+                plugin := pb(plugin.Arguments)
+                if qp, ok := plugin.(QueuePlugin); ok {
+                    qp.InitQueueContext(qCtx, ssn)
+                }
+            }
+        }
+	ssn.QueueContexts[queue.UID] = qCtx
+  
+    return ssn
+}
+```
+
+Since different queues may use the same plugins but with different parameters, we need to create a separate QueueContext for each queue.
+
+## 2.2、action
+
+Currently, the functions registered during the OpenSession phase are triggered in the action, but these functions are globally configured.
+
+Taking the action enqueue as an example:
+
+```go
+// Current action.Execute()
+func (enqueue *Action) Execute(ssn *framework.Session) {
+	klog.V(5).Infof("Enter Enqueue ...")
+	defer klog.V(5).Infof("Leaving Enqueue ...")
+
+	queues := util.NewPriorityQueue(ssn.QueueOrderFn)
+	queueSet := sets.NewString()
+	jobsMap := map[api.QueueID]*util.PriorityQueue{}
+
+	for _, job := range ssn.Jobs {
+		if job.ScheduleStartTimestamp.IsZero() {
+			ssn.Jobs[job.UID].ScheduleStartTimestamp = metav1.Time{
+				Time: time.Now(),
+			}
+		}
+		if queue, found := ssn.Queues[job.Queue]; !found {
+			klog.Errorf("Failed to find Queue <%s> for Job <%s/%s>",
+				job.Queue, job.Namespace, job.Name)
+			continue
+		} else if !queueSet.Has(string(queue.UID)) {
+			klog.V(5).Infof("Added Queue <%s> for Job <%s/%s>",
+				queue.Name, job.Namespace, job.Name)
+
+			queueSet.Insert(string(queue.UID))
+			queues.Push(queue)
+		}
+
+		if job.IsPending() {
+			if _, found := jobsMap[job.Queue]; !found {
+				jobsMap[job.Queue] = util.NewPriorityQueue(ssn.JobOrderFn)
```

Comment:
After you register QueueContext, how do you actually use it? I don't seem to see it. E.g., here it's still `ssn.JobOrderFn`.

Reply:
OK. So you also need to show how to use queue context's extension points in this design doc